File: Others/medium/ThreeNumberSum.py
def threeNumberSum(array, targetSum):
    sols = []
    array.sort()
    for i in range(len(array)):
        target = targetSum - array[i]
        for sol in two_sum(array, target, i):
            sols.append([array[i]] + sol)

    return sols


def two_sum(array, target, curr_index):
    solutions = []

    left = curr_index + 1
    right = len(array) - 1

    while left < right:

        if array[left] + array[right] == target:
            solutions.append([array[left], array[right]])
            left += 1
            right -= 1
            while array[left] == array[left - 1]:
                left += 1
            while array[right] == array[right + 1]:
                right -= 1

        elif array[left] + array[right] < target:
            left += 1
        else:
            right -= 1

    return solutions


print(threeNumberSum([12, 3, 1, 2, -6, 5, -8, 6], 0))

File: Golang/medium/three-number-sum.go
package main

import (
	"sort"
)

type Pair struct {
	first, second int
}

func twoNumberSum(array []int, left int, right int, target int) []Pair {
	result := []Pair{}
	for left < right {
		if array[left]+array[right] == target {
			result = append(result, Pair{array[left], array[right]})
			left += 1
			right -= 1
			for left < len(array)-1 && array[left] == array[left+1] {
				left += 1
			}
			for right > 0 && array[right] == array[right-1] {
				right -= 1
			}
		} else if array[left]+array[right] > target {
			right -= 1
		} else {
			left += 1
		}
	}
	return result
}

func ThreeNumberSum(array []int, target int) [][]int {
	sort.Ints(array)

	res := [][]int{}

	for idx, elem := range array {
		targetSum := target - elem
		sols := twoNumberSum(array, idx+1, len(array)-1, targetSum)

		for _, sol := range sols {
			currSol := []int{elem, sol.first, sol.second}
			res = append(res, currSol)
		}
	}
	return res
}
